File: packages/sdk-py/src/swarmwage/receipts.py
# Swarmwage Agent SDK (Python) — seller-side receipt submission
# Mirrors packages/sdk-ts/src/receipts.ts
# License: MIT
#
# Layer 3 of the 4-layer data-capture model: sellers sign and submit
# receipts after each fulfilled hire so their public reputation
# reflects on-protocol activity. Default ON; opt out via env var
# `SWARMWAGE_RECEIPTS=0`.

from __future__ import annotations

import os
import sys
from dataclasses import dataclass, field
from typing import Any, Callable

import httpx
from eth_account import Account
from eth_keys.exceptions import ValidationError as _KeysValidationError

from ._signing import sign_typed_payload

DEFAULT_REGISTRY_URL = "https://api.swarmwage.com"

# Fields a ReceiptPayload (or equivalent dict) must contain before we'll
# sign it. Missing fields would silently produce a signature over an
# incomplete body that the registry will 4xx — turning a programming bug
# into a confusing fire-and-forget failure. Validate up front instead.
_REQUIRED_RECEIPT_FIELDS = frozenset(
    {
        "protocol_version",
        "hire_id",
        "agent_id",
        "buyer",
        "capability",
        "amount_usdc_atomic",
        "network",
        "tx_hash",
        "completed_at",
        "verification",
    }
)


@dataclass
class ReceiptPayload:
    """The fields the seller signs.

    Sent verbatim to the registry. The signature is computed over the
    canonical-JSON serialization of this dict (see `_signing`).
    """

    protocol_version: str
    hire_id: str
    agent_id: str
    buyer: str
    capability: str
    amount_usdc_atomic: str
    network: str  # "base" | "base-sepolia"
    tx_hash: str
    completed_at: str  # ISO 8601
    verification: dict[str, Any]
    capability_version: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "protocol_version": self.protocol_version,
            "hire_id": self.hire_id,
            "agent_id": self.agent_id,
            "buyer": self.buyer,
            "capability": self.capability,
            "amount_usdc_atomic": self.amount_usdc_atomic,
            "network": self.network,
            "tx_hash": self.tx_hash,
            "completed_at": self.completed_at,
            "verification": self.verification,
        }
        if self.capability_version is not None:
            d["capability_version"] = self.capability_version
        return d


@dataclass
class SubmitReceiptResult:
    submitted: bool
    status: int | None = None
    receipt_id: str | None = None
    error: str | None = None


def is_receipts_enabled(
    enabled: bool | None = None,
    env: dict[str, str] | None = None,
) -> bool:
    """Decide whether receipt submission is enabled.

    Mirrors `AGENT_TELEMETRY` opt-out semantics: env var precedence
    after an explicit `enabled=` flag.
    """
    if enabled is not None:
        return enabled
    env_map = env if env is not None else dict(os.environ)
    raw = env_map.get("SWARMWAGE_RECEIPTS")
    if raw is None:
        return True
    v = raw.strip().lower()
    if v in ("0", "false", "off", "no"):
        return False
    return True


def sign_receipt(
    seller_private_key: str,
    payload: ReceiptPayload | dict[str, Any],
) -> dict[str, Any]:
    """Sign a receipt payload and return the body the registry expects."""
    body = payload.to_dict() if isinstance(payload, ReceiptPayload) else dict(payload)
    signature = sign_typed_payload(seller_private_key, body)
    return {**body, "signature": signature}


def _default_logger(msg: str) -> None:
    try:
        sys.stderr.write(f"swarmwage-receipts: {msg}\n")
    except Exception:
        # logging is best-effort
        pass


def submit_receipt(
    *,
    seller_private_key: str,
    payload: ReceiptPayload | dict[str, Any],
    registry_url: str = DEFAULT_REGISTRY_URL,
    enabled: bool | None = None,
    logger: Callable[[str], None] | None = None,
    http_client: httpx.Client | None = None,
    timeout: float = 10.0,
) -> SubmitReceiptResult:
    """Submit a signed receipt to the registry.

    Fire-and-forget semantics: never raises. Errors are returned in the
    result and logged via `logger`. Safe to call from a seller's `/hire`
    handler without blocking the buyer response.

    Default ON; set env `SWARMWAGE_RECEIPTS=0` to disable.
    """
    log = logger or _default_logger

    if not is_receipts_enabled(enabled):
        return SubmitReceiptResult(submitted=False)

    body_dict = (
        payload.to_dict() if isinstance(payload, ReceiptPayload) else dict(payload)
    )

    # Validate up front so a missing required field surfaces as a clear
    # caller error rather than a "submitted=True, error='sign failed: ...'"
    # ambiguity that masks the bug.
    missing = sorted(_REQUIRED_RECEIPT_FIELDS - set(body_dict.keys()))
    if missing:
        msg = (
            f"receipt payload missing required field(s): {missing}. "
            "Pass a `ReceiptPayload` (dataclass) or a dict with all of: "
            f"{sorted(_REQUIRED_RECEIPT_FIELDS)}."
        )
        log(msg)
        return SubmitReceiptResult(submitted=True, error=msg)

    # Sanity: the payload must declare itself as the wallet we're signing with.
    try:
        wallet_address = Account.from_key(seller_private_key).address.lower()
    except (_KeysValidationError, ValueError, TypeError) as exc:
        # Narrow exception classes so a typo in the key surfaces clearly
        # while real programming errors (AttributeError on a wrong object
        # shape, etc.) bubble up to the caller instead of being swallowed
        # under a generic "sign failed".
        msg = f"sign failed: bad private key ({type(exc).__name__}: {exc})"
        log(msg)
        return SubmitReceiptResult(submitted=True, error=msg)

    payload_agent = str(body_dict.get("agent_id", "")).lower()
    if payload_agent != wallet_address:
        msg = (
            f"payload.agent_id {payload_agent!r} does not match seller wallet "
            f"{wallet_address!r}"
        )
        log(msg)
        return SubmitReceiptResult(submitted=True, error=msg)

    try:
        signed_body = sign_receipt(seller_private_key, body_dict)
    except (_KeysValidationError, ValueError, TypeError) as exc:
        msg = f"sign failed: {type(exc).__name__}: {exc}"
        log(msg)
        return SubmitReceiptResult(submitted=True, error=msg)

    url = f"{registry_url.rstrip('/')}/v1/receipts"
    owns_client = http_client is None
    client = http_client or httpx.Client(timeout=timeout)
    try:
        resp = client.post(url, json=signed_body)
        if resp.status_code >= 400:
            try:
                txt = resp.text
            except Exception:
                txt = ""
            log(f"registry rejected receipt ({resp.status_code}): {txt}")
            return SubmitReceiptResult(
                submitted=True, status=resp.status_code, error=txt or resp.reason_phrase
            )
        try:
            payload_json = resp.json()
        except Exception:
            payload_json = {}
        return SubmitReceiptResult(
            submitted=True,
            status=resp.status_code,
            receipt_id=payload_json.get("receipt_id") if isinstance(payload_json, dict) else None,
        )
    except httpx.HTTPError as exc:
        msg = f"network error: {exc}"
        log(msg)
        return SubmitReceiptResult(submitted=True, error=msg)
    finally:
        if owns_client:
            client.close()


__all__ = [
    "DEFAULT_REGISTRY_URL",
    "ReceiptPayload",
    "SubmitReceiptResult",
    "is_receipts_enabled",
    "sign_receipt",
    "submit_receipt",
]

File: packages/sdk-ts/src/receipts.ts
// Swarmwage Agent SDK — seller-side receipt submission
// Layer 3 of the 4-layer data capture model: sellers sign and submit
// receipts after each fulfilled hire so their public reputation
// reflects on-protocol activity.
// Default ON; opt out via env var SWARMWAGE_RECEIPTS=0.
// License: MIT

import { createWallet, type AgentWallet } from "./wallet.js";
import type {
  AgentId,
  CapabilityId,
  Hex,
  ProtocolVersion,
  VerificationResult,
} from "./types.js";

export const DEFAULT_REGISTRY_URL = "https://api.swarmwage.com";

/**
 * Network the on-chain settlement happened on. Mirrors the SDK's
 * `SwarmwageNetwork` and the registry's accepted enum.
 */
export type ReceiptNetwork = "base" | "base-sepolia";

/**
 * The fields the seller signs. Sent verbatim (alphabetically ordered)
 * to the registry so the server can recover the address.
 */
export interface ReceiptPayload {
  protocol_version: ProtocolVersion | string;
  /** SDK-generated UUID per hire — idempotency key. */
  hire_id: string;
  /** Seller (signer). Lowercase 0x address. */
  agent_id: AgentId;
  /**
   * Settled payment recipient, when distinct from `agent_id` (GH #11).
   * MUST equal the `payee` bound into the seller's signed listing — the same
   * tuple the buyer's anti-hijack check validated. Omit for single-EOA
   * sellers (payments went to `agent_id`).
   */
  payee?: AgentId;
  /** Buyer address (informational). */
  buyer: AgentId;
  capability: CapabilityId;
  capability_version?: string;
  /** USDC atomic units (6 decimals), as a string for BigInt safety. */
  amount_usdc_atomic: string;
  network: ReceiptNetwork;
  tx_hash: `0x${string}`;
  /** ISO 8601 timestamp. */
  completed_at: string;
  verification: VerificationResult | { all_passed: boolean; checks: Record<string, boolean> };
}

export interface SubmitReceiptOptions {
  /** Override the canonical registry URL. Defaults to api.swarmwage.com. */
  registryUrl?: string;
  /** Seller private key, used to sign the payload. */
  sellerPrivateKey: Hex;
  /** Receipt content. Signature is added internally. */
  payload: ReceiptPayload;
  /** Force on/off. Defaults to env var `SWARMWAGE_RECEIPTS` (off=`0|false|off|no`). */
  enabled?: boolean;
  /** Optional logger. Defaults to writing to `process.stderr`. */
  logger?: (msg: string) => void;
  /** Override fetch (for tests). Defaults to global `fetch`. */
  fetchImpl?: typeof fetch;
}

export interface SubmitReceiptResult {
  /** True when the SDK submitted (or attempted to submit). False when opted out. */
  submitted: boolean;
  /** HTTP status code if a response was received. */
  status?: number;
  /** Receipt ID returned by the registry on success. */
  receipt_id?: string;
  /** Error message (if any). Errors are NOT thrown by `submitReceipt`. */
  error?: string;
}

/**
 * Decide whether receipt submission is enabled. Mirrors the
 * `AGENT_TELEMETRY` opt-out semantics for symmetry.
 */
export function isReceiptsEnabled(
  enabled: boolean | undefined,
  env: Record<string, string | undefined> = (typeof process !== "undefined"
    ? (process.env as Record<string, string | undefined>)
    : {}),
): boolean {
  if (enabled !== undefined) return enabled;
  const raw = env.SWARMWAGE_RECEIPTS;
  if (raw === undefined) return true;
  const v = raw.trim().toLowerCase();
  if (v === "0" || v === "false" || v === "off" || v === "no") return false;
  return true;
}

/**
 * Sign a receipt payload using the canonical scheme:
 * stable JSON (alphabetical keys) → keccak256 → eth_personalSign.
 *
 * Returns the `{ signature, body }` so callers can submit elsewhere if
 * they want without going through `submitReceipt`.
 */
export async function signReceipt(
  wallet: AgentWallet,
  payload: ReceiptPayload,
): Promise<{ signature: Hex; body: ReceiptPayload & { signature: Hex } }> {
  const signature = await wallet.signTypedPayload(payload);
  return { signature, body: { ...payload, signature } };
}

/**
 * Submit a signed receipt to the registry. Fire-and-forget semantics:
 * never throws, errors are returned in the result and logged via
 * `opts.logger`. Safe to `await` from a seller's `/hire` handler without
 * blocking the buyer's response.
 *
 * Default ON; set env `SWARMWAGE_RECEIPTS=0` to disable.
 */
export async function submitReceipt(
  opts: SubmitReceiptOptions,
): Promise<SubmitReceiptResult> {
  const log =
    opts.logger ??
    ((msg: string) => {
      try {
        if (typeof process !== "undefined" && process.stderr?.write) {
          process.stderr.write(`swarmwage-receipts: ${msg}\n`);
        }
      } catch {
        // swallow — logging is best-effort
      }
    });

  if (!isReceiptsEnabled(opts.enabled)) {
    return { submitted: false };
  }

  const url = `${(opts.registryUrl ?? DEFAULT_REGISTRY_URL).replace(/\/$/, "")}/v1/receipts`;
  const fetchImpl = opts.fetchImpl ?? globalThis.fetch;

  let signedBody: ReceiptPayload & { signature: Hex };
  try {
    const wallet = createWallet({ privateKey: opts.sellerPrivateKey });
    // Sanity: the payload must declare itself as the wallet we're signing with.
    if (
      opts.payload.agent_id.toLowerCase() !== wallet.agentId.toLowerCase()
    ) {
      const err = `payload.agent_id ${opts.payload.agent_id} does not match seller wallet ${wallet.agentId}`;
      log(err);
      return { submitted: true, error: err };
    }
    const { body } = await signReceipt(wallet, opts.payload);
    signedBody = body;
  } catch (err) {
    const msg = `sign failed: ${(err as Error).message}`;
    log(msg);
    return { submitted: true, error: msg };
  }

  try {
    const res = await fetchImpl(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(signedBody),
      keepalive: true,
    });
    if (!res.ok) {
      const txt = await res.text().catch(() => "");
      log(`registry rejected receipt (${res.status}): ${txt}`);
      return { submitted: true, status: res.status, error: txt || res.statusText };
    }
    const json = (await res.json().catch(() => ({}))) as {
      receipt_id?: string;
      ok?: boolean;
    };
    return {
      submitted: true,
      status: res.status,
      receipt_id: json.receipt_id,
    };
  } catch (err) {
    const msg = `network error: ${(err as Error).message}`;
    log(msg);
    return { submitted: true, error: msg };
  }
}
